### PR TITLE
Migrate to ubuntu-24.04 runner 

### PR DESCRIPTION
**Pull request by andife**
### Description

The g++ default is g++13 for ubuntu 24.04. 
Use g++ 11 in main.yml until we decide to use a higher g++ version.

(g++13 needs patching some protobuf files, or we also upgrade protobuf...)

### CODE REVIEW

**Comment by andife** [comment]
@justinchuby @cyyever 

difference of runners are explained here: https://github.com/actions/runner-images/issues/10636

One difference is that  
 --   C++ compiler version              : 13.2.0 is used
instead of 
  --   C++ compiler version              : 11.4.0


ubuntu 24.04: https://github.com/onnx/onnx/actions/runs/11156019248/job/31007898431?pr=6373
ubuntu 22.XX/latest: https://github.com/onnx/onnx/actions/runs/11156022942/job/31007907821?pr=6370

**Comment by cyyever** [comment]
The CI failures are due to protobuf lacking a header inclusion, one solution is to patch the downloaded source code in CMake.

**Comment by andife** [comment]
https://github.com/google/mozc/issues/742

**Comment by andife** [comment]
Should we patch the files, that the headers are included? Select g++ 12 in the hope that this still fits? (is installed at ubuntu 24.04) Or increase the protobuf version?

Maybe it relates to https://github.com/onnx/onnx/issues/6298 ?

@cyyever @justinchuby @xadupre 

The support for ubuntu 22.04 is up to Apr 2027 ( https://ubuntu.com/about/release-cycle)
I'll create a PR replacing ubuntu-latest with ubuntu-22.04

**Comment by cyyever** [comment]
@andife It's better patch it now.

**Comment by justinchuby** [comment]
Bumping protobuf version sounds good to me. Thanks!

**Comment by cyyever** [comment]
The failure is interesting. How can macos run "apt-get"？

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -98,9 +105,10 @@ jobs:
           python -m pip install protobuf==3.20.2
           git submodule update --init --recursive
 
+
       - name: Build and install ONNX - Linux
-        if: matrix.os == 'ubuntu-latest'
-        run: |
+        if: (matrix.os == 'ubuntu-latest' || matrix.os == 'ubuntu-24.04')
+        run: |
```
Comment (cyyever) [inline .github/workflows/main.yml:111]: Empty space behind ‘|’

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -151,7 +159,7 @@ jobs:
           pytest -sv --cov=onnx --cov-report=xml --cov-append --cov-branch --junitxml junit.xml -n auto --dist loadscope
 
       - name: Run C++ tests
-        if: matrix.os == 'ubuntu-latest' || matrix.os == 'macos-latest'
+        if: (matrix.os == 'ubuntu-latest' || matrix.os == '24.04') || matrix.os == 'macos-latest'
```
Comment (cyyever) [inline .github/workflows/main.yml:162]: Parentheses are not need

**Comment by cyyever** [comment]
After the PR get merged is it possible to upgrade protobuf?

**Comment by andife** [comment]
ready to review

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -70,8 +70,15 @@ jobs:
...
+      - name: Show versions g++
+        if: (matrix.os == 'ubuntu-latest' || matrix.os == 'ubuntu-24.04')
+        run: |
+          sudo apt install g++-11
+          sudo update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-11 12
+          g++ --version          
+
       - name: Install external protobuf - Linux
-        if: matrix.protobuf_type == 'External' && matrix.os == 'ubuntu-latest'
+        if: matrix.protobuf_type == 'External' && (matrix.os == 'ubuntu-latest' || matrix.os == 'ubuntu-24.04')
```
Comment (justinchuby) [inline .github/workflows/main.yml:81]: Startswith ubuntu?

Comment (andife) [reply .github/workflows/main.yml:81]: yes, thank you

**Comment by justinchuby** [comment]
Is it possible to just upgrade protobuf? I would prefer that over patching the source. It seems we have time before Apr 2027

**Comment by cyyever** [comment]
Have a working ubuntu-24.04 example would make upgrade less risky.

**Comment by justinchuby** [comment]
I don't think there is too much risk with that? We can always revert if things do not work

**Comment by andife** [comment]
Regarding the ubuntu 24.04 pr... of course we could stay longer on our 22.04 version, because it is supported longer. From the Ubuntu side. Then we only have to change ubuntu-latest to ubuntu-22.04

Protobuf is not be patched in the PR. Only the g++ version we have been using so far will be installed and used (or is that what you mean by patching?). For me, this is a smaller intervention than pulling up the protobuf version?

File: .github/workflows/main.yml (hunk 3 above)
Comment (justinchuby) [reply .github/workflows/main.yml:81]: I think you may use https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/evaluate-expressions-in-workflows-and-actions#startswith to simplify this expression

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -201,8 +209,15 @@ jobs:
...
+        if: (matrix.python_version == '3.11' || matrix.python_version == '3.12') &&  (matrix.os == 'windows-latest' || matrix.os == 'macos-latest')
+        run: |
+          pip install "numpy<2.0" pillow
+          pytest -s
+
+      - name: Run Python tests with numpy<2.0 (ubuntu)
+        if: (matrix.python_version == '3.11' || matrix.python_version == '3.12' ) && (matrix.os == 'ubuntu-latest' || matrix.os == 'ubuntu-24.04')
         run: |
-          pip install "numpy<2.0" "pillow<10"
+          sudo apt-get install libjpeg-dev zlib1g-dev libpng-dev libtiff-dev libfreetype6-dev
```
Comment (justinchuby) [inline .github/workflows/main.yml:221]: Could you add a comment on why this is needed?

File: .github/workflows/main.yml [hunk 5]
```diff
@@ -70,8 +70,15 @@ jobs:
           python --version
           cmake --version
 
+      - name: Show versions g++
+        if: (matrix.os == 'ubuntu-latest' || matrix.os == 'ubuntu-24.04')
+        run: |
+          sudo apt install g++-11
```
Comment (justinchuby) [inline .github/workflows/main.yml:76]: Possible to add a todo to remove this?

**Comment by justinchuby** [comment]
> Regarding the ubuntu 24.04 pr... of course we could stay longer on our 22.04 version, because it is supported longer. From the Ubuntu side. Then we only have to change ubuntu-latest to ubuntu-22.04
> 
> Protobuf is not be patched in the PR. Only the g++ version we have been using so far will be installed and used (or is that what you mean by patching?). For me, this is a smaller intervention than pulling up the protobuf version?

Sounds good. Previously I saw the protobuf code was patch but apparently this is no longer the case. I added some comments to simplify the change. Would be helpful to add comments in the yml file to explain some of the changes for future readers too. Otherwise looks good to me, thanks!

**Review by justinchuby — APPROVED** [review]

**Comment by justinchuby** [comment]
So the `Test ubuntu-latest` workflows are required. We will need to change those too.

File: .github/workflows/main.yml (hunk 5 above)
Comment (andife) [reply .github/workflows/main.yml:78]: added

File: .github/workflows/main.yml [hunk 6]
```diff
@@ -216,8 +218,9 @@ jobs:
           pytest -s
 
       - name: Run Python tests with numpy<2.0 (ubuntu)
-        if: (matrix.python_version == '3.11' || matrix.python_version == '3.12' ) && (matrix.os == 'ubuntu-latest' || matrix.os == 'ubuntu-24.04')
+        if: (matrix.python_version == '3.11' || matrix.python_version == '3.12' ) && startsWith(matrix.os,'ubuntu')
         run: |
-          sudo apt-get install libjpeg-dev zlib1g-dev libpng-dev libtiff-dev libfreetype6-dev
+          # 2024.10.15: Error message: The headers or library files could not be found for jpeg, a required dependency when compiling Pillow from source.
+          sudo apt-get install libjpeg-dev zlib1g-dev libpng-dev
```
Comment (justinchuby) [inline .github/workflows/main.yml:224]: Why do we need to compile pillow?

Comment (andife) [reply .github/workflows/main.yml:224]: The following runs fails: https://github.com/onnx/onnx/actions/runs/11306051777/job/31446447690
We want to have numpy < 2, which means pillow<10, so we have pillow 9.5.0 and that version is not supported by python 3.12 (https://pillow.readthedocs.io/en/latest/installation/python-support.html#python-support). Might work with python 3.11? I'm not sure why this was not a problem with ubuntu 22.04, was there a corresponding dependency already installed elsewhere?

Comment (justinchuby) [reply .github/workflows/main.yml:224]: I guess we will want numpy2 support first then?

Comment (justinchuby) [reply .github/workflows/main.yml:224]: I personally don’t think we should compile any dependencies

Comment (andife) [reply .github/workflows/main.yml:224]: > I guess we will want numpy2 support first then?

We have already numpy 2 support. Here at that step we make sure that everything also works with older numpy versions

Comment (andife) [reply .github/workflows/main.yml:224]: > I personally don’t think we should compile any dependencies

I'll have a look

Comment (andife) [reply .github/workflows/main.yml]: also our current pipeline with ubuntu-22.04 seems to build pillow: 
https://github.com/onnx/onnx/actions/runs/11352377878/job/31574997465#step:21:5845

File: .github/workflows/main.yml (hunk 4 above)
Comment (andife) [reply .github/workflows/main.yml]: added

File: .github/workflows/main.yml (hunk 3 above)
Comment (andife) [reply .github/workflows/main.yml]: adapted

**Review by cyyever — APPROVED** [review]

**Comment by andife** [comment]
@justinchuby Do we use https://docs.github.com/en/enterprise-server@3.11/actions/sharing-automations/required-workflows ?

**Comment by andife** [comment]
@liqunfu @justinchuby I adapted the branch protection rules

**Pull request merged**